IPython/core/events.py

```diff
@@ -56,11 +78,24 @@ def register(self, event, function):
         """
         if not callable(function):
             raise TypeError('Need a callable, got %r' % function)
-        self.callbacks[event].append(function)
+
+        callback_proto = available_events.get(event)
+        if (isfunction(callback_proto) and isfunction(function) and
+            len(getfullargspec(callback_proto).args) > 0 and
+            len(getfullargspec(function).args) == 0):
+            # `callback_proto` has args but `function` does not, so a
+            # compatibility wrapper is needed.
+            self.callbacks[event].append(_compatibility_wrapper_for(function))
```

Comment:
Is there a good reason to implement this functionality within IPython rather than adding a dependency on backcall and using that? We can improve backcall if needed, but it's more flexible than this implementation - e.g. you can have a callback that gets only some of the arguments available.

Reply:
I didn't know that adding `backcall` as a dependency is an option and that you'd prefer that. I just wanted to replicate the behavior we needed for this feature here without adding to much overhead. Will update the PR again when I have some time...